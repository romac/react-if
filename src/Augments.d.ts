declare namespace NodeJS {
  export interface ProcessEnv {
    NODE_ENV: 'development' | 'production' | 'test';
  }
}
