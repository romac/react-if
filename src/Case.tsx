import { FC } from 'react';
import { render } from './render';
import { ComponentWithConditionProps } from './types';

/** If the `<Case />` is the first one to have its condition evaluates to true inside the parent `<Switch />` it will be the only rendered. */
export const Case: FC<ComponentWithConditionProps> = props => render(props);

Case.defaultProps = {
  children: null
};
