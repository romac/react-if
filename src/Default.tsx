import { FC } from 'react';
import { render } from './render';

/** If no `<Case />` have its condition evaluates to true inside the parent `<Switch />`, the first `<Default />` will be the only one rendered. */
export const Default: FC = props => render(props);

Default.defaultProps = {
  children: null
};
