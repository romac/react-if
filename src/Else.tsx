import { render } from './render';
import type { FCWithImplicitChildren } from './types';

/**
 * Must only contain a single child, which it renders as-is.
 * Should not be used outside of an `<If />` block.
 * @param props The props to pass down to the `<Else />` component
 */
export const Else: FCWithImplicitChildren = (props) => render(props);
