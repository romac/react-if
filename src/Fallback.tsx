import { render } from './render';
import type { FCWithImplicitChildren } from './types';

/**
 * Must contain only a single child, which it renders as-is.
 * Should not be used outside of an `<If />` block whose condition prop is a promise.
 * @param props The props to pass down to the `<Fallback />` component
 */
export const Fallback: FCWithImplicitChildren = (props) => render(props);
