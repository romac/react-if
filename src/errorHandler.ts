export const handleError = (errorMessage: string) => {
  if (process.env.NODE_ENV !== 'production') {
    throw new Error(errorMessage);
  }
};
