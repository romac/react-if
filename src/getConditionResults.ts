import { BooleanLike } from './types';

export const getConditionResult = (condition: BooleanLike | ((...args: unknown[]) => BooleanLike)) => {
  const conditionResult = Boolean(typeof condition === 'function' ? condition() : condition);

  return conditionResult;
};
