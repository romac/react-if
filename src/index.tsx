export * from './Case';
export * from './Default';
export * from './Else';
export * from './If';
export * from './Switch';
export * from './Then';
export * from './types';
export * from './Unless';
export * from './When';
