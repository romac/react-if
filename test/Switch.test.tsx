import { mount, shallow } from 'enzyme';
import React from 'react';
import { Case, Default, Switch } from '../src';

describe('<Switch /> component', () => {
  describe('Renders children <Case />', () => {
    test('GIVEN multiple cases w/ truthy condition THEN renders children of first <Case />', () => {
      const wrapped = shallow(
        <Switch>
          <Case condition={true}>
            <span>Case1</span>
          </Case>
          <Case condition={true}>
            <span>Case2</span>
          </Case>
        </Switch>
      );

      expect(wrapped).toMatchSnapshot();
      expect(wrapped.containsMatchingElement(<span>Case1</span>)).toBe(true);
      expect(wrapped.containsMatchingElement(<span>Case2</span>)).toBe(false);
    });

    test('GIVEN multiple cases w/ falsy condition then truthy condition THEN renders children of second <Case />', () => {
      const wrapped = shallow(
        <Switch>
          <Case condition={false}>
            <span>Case1</span>
          </Case>
          <Case condition={true}>
            <span>Case2</span>
          </Case>
        </Switch>
      );

      expect(wrapped).toMatchSnapshot();
      expect(wrapped.containsMatchingElement(<span>Case1</span>)).toBe(false);
      expect(wrapped.containsMatchingElement(<span>Case2</span>)).toBe(true);
    });

    test('GIVEN multiple cases w/ falsy condition then function as truthy condition THEN renders children of second <Case />', () => {
      const wrapped = shallow(
        <Switch>
          <Case condition={false}>
            <span>Case1</span>
          </Case>
          <Case condition={() => true}>
            <span>Case2</span>
          </Case>
        </Switch>
      );

      expect(wrapped).toMatchSnapshot();
      expect(wrapped.containsMatchingElement(<span>Case1</span>)).toBe(false);
      expect(wrapped.containsMatchingElement(<span>Case2</span>)).toBe(true);
    });

    test('GIVEN condition as function THEN resolves for first <Case />', () => {
      const wrapped = shallow(
        <Switch>
          <Case condition={() => true}>
            <span>Case1</span>
          </Case>
          <Case condition={() => false}>
            <span>Case2</span>
          </Case>
          <Case condition={() => true}>
            <span>Case3</span>
          </Case>
        </Switch>
      );

      expect(wrapped).toMatchSnapshot();
      expect(wrapped.containsMatchingElement(<span>Case1</span>)).toBe(true);
      expect(wrapped.containsMatchingElement(<span>Case2</span>)).toBe(false);
      expect(wrapped.containsMatchingElement(<span>Case3</span>)).toBe(false);
    });

    test('GIVEN child as function THEN resolves for first <Case />', () => {
      const wrapped = mount(
        <Switch>
          <Case condition={() => true}>{() => <span>Case1</span>}</Case>
          <Case condition={() => false}>
            <span>Case2</span>
          </Case>
          <Case condition={() => true}>
            <span>Case3</span>
          </Case>
        </Switch>
      );

      expect(wrapped).toMatchSnapshot();
      expect(wrapped.containsMatchingElement(<span>Case1</span>)).toBe(true);
      expect(wrapped.containsMatchingElement(<span>Case2</span>)).toBe(false);
      expect(wrapped.containsMatchingElement(<span>Case3</span>)).toBe(false);
    });
  });

  describe('Renders the child element of <Default />', () => {
    test('GIVEN simple <Default /> THEN renders children of <Default />', () => {
      const wrapped = shallow(
        <Switch>
          <Default>
            <span>Default</span>
          </Default>
        </Switch>
      );

      expect(wrapped).toMatchSnapshot();
      expect(wrapped.containsMatchingElement(<span>Default</span>)).toBe(true);
    });

    test('GIVEN falsy <Case /> and <Default /> THEN renders children of <Default />', () => {
      const wrapped = shallow(
        <Switch>
          <Case condition={false}>
            <span>Case</span>
          </Case>
          <Default>
            <span>Default</span>
          </Default>
        </Switch>
      );

      expect(wrapped).toMatchSnapshot();
      expect(wrapped.containsMatchingElement(<span>Default</span>)).toBe(true);
      expect(wrapped.containsMatchingElement(<span>Case</span>)).toBe(false);
    });

    test('GIVEN truthy <Case /> and <Default /> THEN renders children of <Case />', () => {
      const wrapped = shallow(
        <Switch>
          <Case condition={true}>
            <span>Case</span>
          </Case>
          <Default>
            <span>Default</span>
          </Default>
        </Switch>
      );

      expect(wrapped).toMatchSnapshot();
      expect(wrapped.containsMatchingElement(<span>Case</span>)).toBe(true);
      expect(wrapped.containsMatchingElement(<span>Default</span>)).toBe(false);
    });

    test('GIVEN children as function THEN renders children of <Default />', () => {
      const wrapped = mount(
        <Switch>
          <Case condition={false}>{() => <span>Case</span>}</Case>
          <Default>{() => <span>Default</span>}</Default>
        </Switch>
      );

      expect(wrapped).toMatchSnapshot();
      expect(wrapped.containsMatchingElement(<span>Default</span>)).toBe(true);
      expect(wrapped.containsMatchingElement(<span>Case</span>)).toBe(false);
    });
  });

  describe('Renders nothing in some cases', () => {
    test('GIVEN no children THEN renders nothing', () => {
      const wrapped = shallow(<Switch></Switch>);

      expect(wrapped).toMatchSnapshot();
      expect(wrapped.html()).toBeNull();
    });

    test('GIVEN multiple cases w/ all falsy condition THEN renders nothing', () => {
      const wrapped = shallow(
        <Switch>
          <Case condition={false}>
            <span>Case1</span>
          </Case>
          <Case condition={false}>
            <span>Case2</span>
          </Case>
        </Switch>
      );

      expect(wrapped).toMatchSnapshot();
      expect(wrapped.containsMatchingElement(<span>Case1</span>)).toBe(false);
      expect(wrapped.containsMatchingElement(<span>Case2</span>)).toBe(false);
      expect(wrapped.html()).toBeNull();
    });

    test('GIVEN invalid React children THEN does not render those', () => {
      const wrapped = shallow(
        <Switch>
          {() => [{ key: 'one', prop: 'two' }]}
          <Case condition={true}>
            <span>Case2</span>
          </Case>
        </Switch>
      );

      expect(wrapped).toMatchSnapshot();
      expect(wrapped.containsMatchingElement(<span>Case1</span>)).toBe(false);
      expect(wrapped.containsMatchingElement(<span>Case2</span>)).toBe(true);
    });
  });
});
