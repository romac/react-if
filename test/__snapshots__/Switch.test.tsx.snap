// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch /> component Renders children <Case /> GIVEN child as function THEN resolves for first <Case /> 1`] = `
<span>
  Case1
</span>
`;

exports[`<Switch /> component Renders children <Case /> GIVEN condition as function THEN resolves for first <Case /> 1`] = `
<Component
  condition={[Function]}
>
  <span>
    Case1
  </span>
</Component>
`;

exports[`<Switch /> component Renders children <Case /> GIVEN multiple cases w/ falsy condition then function as truthy condition THEN renders children of second <Case /> 1`] = `
<Component
  condition={[Function]}
>
  <span>
    Case2
  </span>
</Component>
`;

exports[`<Switch /> component Renders children <Case /> GIVEN multiple cases w/ falsy condition then truthy condition THEN renders children of second <Case /> 1`] = `
<Component
  condition={true}
>
  <span>
    Case2
  </span>
</Component>
`;

exports[`<Switch /> component Renders children <Case /> GIVEN multiple cases w/ truthy condition THEN renders children of first <Case /> 1`] = `
<Component
  condition={true}
>
  <span>
    Case1
  </span>
</Component>
`;

exports[`<Switch /> component Renders nothing in some cases GIVEN invalid React children THEN does not render those 1`] = `
<Component
  condition={true}
>
  <span>
    Case2
  </span>
</Component>
`;

exports[`<Switch /> component Renders nothing in some cases GIVEN multiple cases w/ all falsy condition THEN renders nothing 1`] = `""`;

exports[`<Switch /> component Renders nothing in some cases GIVEN no children THEN renders nothing 1`] = `""`;

exports[`<Switch /> component Renders the child element of <Default /> GIVEN children as function THEN renders children of <Default /> 1`] = `
<span>
  Default
</span>
`;

exports[`<Switch /> component Renders the child element of <Default /> GIVEN falsy <Case /> and <Default /> THEN renders children of <Default /> 1`] = `
<Component>
  <span>
    Default
  </span>
</Component>
`;

exports[`<Switch /> component Renders the child element of <Default /> GIVEN simple <Default /> THEN renders children of <Default /> 1`] = `
<Component>
  <span>
    Default
  </span>
</Component>
`;

exports[`<Switch /> component Renders the child element of <Default /> GIVEN truthy <Case /> and <Default /> THEN renders children of <Case /> 1`] = `
<Component
  condition={true}
>
  <span>
    Case
  </span>
</Component>
`;
